projects/driver_controls/src/drive_output.c

```diff
@@ -0,0 +1,90 @@
+#include "drive_output.h"
+#include <string.h>
+#include "log.h"
+
+#define DRIVE_OUTPUT_VALID_WATCHDOG ((1 << NUM_DRIVE_OUTPUT_SOURCES) - 1)
+
+static DriveOutputStorage s_storage;
+
+static void prv_watchdog_cb(SoftTimerID timer_id, void *context) {
+  DriveOutputStorage *storage = context;
+
+  // We're missing at least one updated response
+  if (storage->watchdog != DRIVE_OUTPUT_VALID_WATCHDOG) {
+    // Error - raise a warning, disable periodic drive storage
+    soft_timer_cancel(storage->output_timer);
+    event_raise(storage->fault_event, 0);
+  }
+
+  // Reset watchdog
+  storage->watchdog = 0;
+
+  soft_timer_start_millis(DRIVE_OUTPUT_WATCHDOG_MS, prv_watchdog_cb, context,
+                          &storage->watchdog_timer);
+}
+
+static void prv_broadcast_cb(SoftTimerID timer_id, void *context) {
+  DriveOutputStorage *storage = context;
+
+  event_raise(storage->update_req_event, 0);
+  // note that this will usually output stale data from the previous update request
+
+  // TODO(ELEC-354): output CAN message
+  LOG_DEBUG("Drive output: throttle %d cruise %d direction %d steering angle %d\n",
+            storage->data[DRIVE_OUTPUT_SOURCE_THROTTLE], storage->data[DRIVE_OUTPUT_SOURCE_CRUISE],
+            storage->data[DRIVE_OUTPUT_SOURCE_DIRECTION],
+            storage->data[DRIVE_OUTPUT_SOURCE_STEERING_ANGLE]);
+
+  soft_timer_start_millis(DRIVE_OUTPUT_BROADCAST_MS, prv_broadcast_cb, context,
+                          &storage->output_timer);
+}
+
+StatusCode drive_output_init(DriveOutputStorage *storage, EventID fault_event,
+                             EventID update_req_event) {
+  memset(storage, 0, sizeof(*storage));
+  storage->fault_event = fault_event;
+  storage->update_req_event = update_req_event;
+
+  storage->watchdog_timer = SOFT_TIMER_INVALID_TIMER;
```

Comment:
Should cancel any still running timers?

Reply:
Hmm, so one concern I've realized with that is that we're relying on the storage provided to have some "valid" or "cleared" state. If we pass in arbitrary storage that isn't cleared, we can't ensure that the stored timer is actually a valid timer to cancel.

Reply:
Hmm, that's a fair point. Perhaps you could add some kind of destroy function which properly deconstructs the struct instead then?

This assumption is made in a lot of places so we should probably file a ticket to audit and fix it elsewhere?

Reply:
I think the idea was that inits should reset everything, since "destroy" is only used for unit testing.

Also, I just realized that since we should reinit soft timers, all running timers should actually be cancelled by the time this function runs.

Reply:
True. Ok sounds fine to me. I think this is really only an issue when testing but since we re-init all the dependencies I get your point. That must be why we never ran into this before.